csharp/src/Apache.Arrow.Adbc/Tracing/TracingBase.cs

```diff
@@ -0,0 +1,351 @@
+﻿/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow.Adbc.Tracing
+{
+    /// <summary>
+    /// Provides a base implementation for a tracing source. If drivers want to enable tracing,
+    /// they need to add a trace listener (e.g., <see cref="FileExporter"/>).
+    /// </summary>
+    public class TracingBase : IDisposable
+    {
+        internal const string ProductVersionDefault = "1.0.0";
+        private static readonly string s_assemblyVersion = GetProductVersion();
+        private bool _disposedValue;
+
+        /// <summary>
+        /// Constructs a new <see cref="TracingBase"/> object. If <paramref name="activitySoureceName"/> is set, it provides the
+        /// activity source name, otherwise the current assembly name is used as the acctivity source name.
+        /// </summary>
+        /// <param name="activitySoureceName"></param>
+        protected TracingBase(string? activitySoureceName = default)
+        {
+            activitySoureceName ??= GetType().Assembly.GetName().Name!;
+            if (string.IsNullOrWhiteSpace(activitySoureceName))
+            {
+                throw new ArgumentNullException(nameof(activitySoureceName));
+            }
+
+            // This is required to be disposed
+            ActivitySource = new(activitySoureceName, s_assemblyVersion);
+        }
+
+        /// <summary>
+        /// Gets the <see cref="ActivitySource"/>.
+        /// </summary>
+        protected ActivitySource ActivitySource { get; }
+
+        /// <summary>
+        /// Gets the name of the <see cref="System.Diagnostics.ActivitySource"/>
+        /// </summary>
+        public string ActivitySourceName => ActivitySource.Name;
```

Comment:
This is an update to the public API. Is this allowed? Or will we need to stage this at a later time?

Reply:
An alternative is to take a Driver/Connection property that sets the name of the `ActivitySource`.